### PR TITLE
[8.x] Fix not found records when using implicit model binding and UUID's

### PR DESCRIPTION
**Pull request by fcolecumberri**
My fix proposal I shared here https://github.com/laravel/framework/issues/37549 and @driesvints asked me to do the PR.

### CODE REVIEW

File: src/Illuminate/Database/Eloquent/Model.php [hunk 1]
```diff
@@ -1792,7 +1792,12 @@ public function getRouteKeyName()
      */
     public function resolveRouteBinding($value, $field = null)
     {
-        return $this->where($field ?? $this->getRouteKeyName(), $value)->first();
+        try {
+            return $this->where($field ?? $this->getRouteKeyName(), $value)->first();
+        } catch (QueryException $e) {
+            // URL type failed to cast into the DB type
+            return null;
+        }
```
Comment (crynobone) [inline src/Illuminate/Database/Eloquent/Model.php:1800]:
```suggestion
        rescue(function () use ($field, value) {
            return $this->where($field ?? $this->getRouteKeyName(), $value)->first();
        }, null);
```

This should be sufficient.

Comment (driesvints) [reply src/Illuminate/Database/Eloquent/Model.php:1800]: @crynobone don't you mean to return the value from `rescue`?

Comment (crynobone) [reply src/Illuminate/Database/Eloquent/Model.php:1800]: Sorry, missed that out

Comment (driesvints) [reply src/Illuminate/Database/Eloquent/Model.php:1800]: @fcolecumberri Can you update the pr with Mior's proposal?

Comment (fcolecumberri) [reply src/Illuminate/Database/Eloquent/Model.php]: Done

File: src/Illuminate/Database/Eloquent/Model.php [hunk 2]
```diff
@@ -1792,7 +1792,9 @@ public function getRouteKeyName()
      */
     public function resolveRouteBinding($value, $field = null)
     {
-        return $this->where($field ?? $this->getRouteKeyName(), $value)->first();
+        return rescue(function () use ($field, value) {
```
Comment (crynobone) [inline src/Illuminate/Database/Eloquent/Model.php:1795]:
```suggestion
        return rescue(function () use ($field, $value) {
```

Sorry, one typo.

Comment (fcolecumberri) [reply src/Illuminate/Database/Eloquent/Model.php]: > Sorry, one typo.

LOL all this PR is about fixing typos

**Comment by BrandonSurowiec** [comment]
The `rescue()` helper by default has `$report = true` as the third parameter. Is this an exception that should be reported by the framework?

**Comment by fcolecumberri** [comment]
> The `rescue()` helper by default has `$report = true` as the third parameter. Is this an exception that should be reported by the framework?

If the exception is thrown because of a typo, IMHO, no, by any mean. If the exception is thrown by another thing (like an error connecting with the DB), you might want to have that reported.

**Comment by taylorotwell** [comment]
I'm afraid this could hide legitimate errors during development that we may want to know about and could be confusing to debug.

You could always override this method on a base model of your application to customize its behavior.

**Comment by fcolecumberri** [comment]
what about a condition that checks if the environment is on debug mode? I still don't like the idea of a typo generating a 500 error.

**Comment by taylorotwell** [comment]
Also - just do this?

```php
Route::get('/users/{id}', ...)->whereUuid('id');
```

**Comment by fcolecumberri** [comment]
Ok, as a short patch I think that works, but in the future I'll make a PR that allow whereUuid to work with model binding.

**Comment by fcolecumberri** [comment]
Ok, I just realized it also works with model binding, I need to do
```php
Route::get('/users/{user:id}', ...)->whereUuid('user');
```

**Pull request closed without merging**